### PR TITLE
Provision per-app gh resources

### LINKED ISSUE #72: Define per-app GitHub configuration

### LINKED ISSUE #73: Provision repo and app contributors GH team with write access
@jedheaj314 
- [x] new repository with app name and description
- [x] new GitHub team with all users defined in `contributors` block
- [x] Link new GH team as contributors to the new repo
- [x] create a deployment branch with the environment name from core config
- [x] add branch protection policy on this branch as per the `branch_approvers` config

@jjgriff93 
- [x] create GH environment with same name as FlowEHR environment
- [x] Create repository scoped token in `platform.tf` and use in `repository.tf` (https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/container_registry_token)
- [x] Add ACR repository token secret as environment secret
- [x] Link the gh environment to only be deployable to using deployment branch

### LINKED ISSUE #74: Provision environment, env-linked secret and protected deployment branch

### PR DESCRIPTION
**Pull request by jjgriff93**
Resolves #73 
Resolves #74 
Resolves #72 

This PR creates all the required resources for an app being hosted within a managed GitHub repo on `make apps`

Consists of:
- [x] new repository with app name and description
- [x] new GitHub team with all users defined in `contributors` block
- [x] Link new GH team as contributors to the new repo
- [x] create a deployment branch with the environment name from core config
- [x] add branch protection policy on this branch as per the `branch_approvers` config
- [x] create GH environment with same name as FlowEHR environment
- [x] Create repository scoped token in `platform.tf` and use in `repository.tf` (https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/container_registry_token)
- [x] Add ACR repository token secret as environment secret
- [x] Link the gh environment to only be deployable to using deployment branch

### CODE REVIEW

File: apps/app/repository.tf [hunk 1]
```diff
@@ -12,4 +12,162 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 
-// TODO: GH stuff will go here
+resource "github_repository" "app" {
+  name        = var.app_id
```
Comment (t-young31) [inline apps/app/repository.tf:16]: I guess this is obvious in TRE land but `app_id` feels like it should be a uid or a azure resource id, to me at least. I'd prefer `app_name`

File: apps/app/repository.tf [hunk 2]
```diff
@@ -12,4 +12,162 @@
...
+
+resource "github_branch_protection" "deployment" {
+  repository_id       = github_repository.app.name
+  pattern             = var.environment
+  allows_deletions    = false
+  allows_force_pushes = false
+
+  required_status_checks {
+    strict   = true
+    contexts = ["lint"]
```
Comment (t-young31) [inline apps/app/repository.tf:98]:
```suggestion
    contexts = ["Lint"]
```
https://github.com/UCLH-Foundry/Dash-Seedling/blob/1191b390e0ff82aa42ca21d31e173f3d90d026cc/.github/workflows/lint.yml#L1

File: apps/app/repository.tf [hunk 3]
```diff
@@ -12,4 +12,162 @@
...
+  repository  = github_repository.app.name
+  environment = var.environment
+
+  deployment_branch_policy {
+    custom_branch_policies = true
+    protected_branches     = false
+  }
+
+  # TODO: remove when https://github.com/integrations/terraform-provider-github/pull/1530 is merged
+  provisioner "local-exec" {
```
Comment (t-young31) [inline apps/app/repository.tf:142]: fun times

File: apps/apps.sample.yaml [hunk 4]
```diff
@@ -2,13 +2,14 @@
 dash_seedling: # unique ID for the app
   name: Dash Seedling
   description: Hello world app for Dash
-  owner: me@mycompany.com
+  owners: # owners to add to repo (for reviews)
```
Comment (t-young31) [inline apps/apps.sample.yaml:5]:
```suggestion
  owners: # GitHub usernames of owners to add to repo (for reviews)
```

File: terragrunt.hcl [hunk 5]
```diff
@@ -117,6 +125,7 @@ inputs = {
   location = get_env("LOCATION")
   naming_suffix = get_env("NAMING_SUFFIX")
   truncated_naming_suffix = get_env("TRUNCATED_NAMING_SUFFIX")
+  environment = get_env("ENVIRONMENT")
```
Comment (t-young31) [inline terragrunt.hcl:128]: I guess this maybe comes into the apps deploy on CI story but I imagine tf will be unhappy with deploying a repo that already exists that doesn't exist in the unshared state between environments. Wonder if it's easier to have a `count` on the resource(s) that are shared across environments now (i.e. if exists count = 0)?

**Review by t-young31 — APPROVED** [review]
Awesome 🌊

File: apps/app/repository.tf (hunk 1 above)
Comment (jjgriff93) [reply apps/app/repository.tf:16]: It should be a uid as it needs to be unique - there's another field for a friendly app name

Comment (jjgriff93) [reply apps/app/repository.tf:16]: Or are you saying just for this resource we should use the app name instead? There's a danger of conflicts if two apps have the same friendly name

File: apps/app/repository.tf (hunk 3 above)
Comment (jjgriff93) [reply apps/app/repository.tf:142]: Yes, "fun"

File: apps/app/repository.tf (hunk 1 above)
Comment (t-young31) [reply apps/app/repository.tf:16]: yeah sorry meant like a guid. ignore me. think it's clear

**Review by jedheaj314 — APPROVED** [review]

File: .gitignore [hunk 6]
```diff
@@ -1,6 +1,5 @@
 # Local .terraform directories
 **/.terraform/*
-.terraform.lock.hcl
```
Comment (jedheaj314) [inline .gitignore:3]: Why have we removed this from gitignore?

**Pull request merged**